File: backend/src/domain/services/auth_service.py
from datetime import datetime, timedelta, timezone
from typing import Optional

import bcrypt
from jose import JWTError, jwt

from src.config import settings


class AuthService:
    """认证服务 - 处理密码哈希和JWT令牌"""

    @staticmethod
    def hash_password(password: str) -> str:
        """对密码进行哈希处理"""
        password_bytes = password.encode('utf-8')
        salt = bcrypt.gensalt()
        hashed = bcrypt.hashpw(password_bytes, salt)
        return hashed.decode('utf-8')

    @staticmethod
    def verify_password(plain_password: str, hashed_password: str) -> bool:
        """验证密码是否匹配"""
        password_bytes = plain_password.encode('utf-8')
        hashed_bytes = hashed_password.encode('utf-8')
        return bcrypt.checkpw(password_bytes, hashed_bytes)

    @staticmethod
    def create_access_token(user_id: str, expires_delta: Optional[timedelta] = None) -> str:
        """创建JWT访问令牌"""
        if expires_delta:
            expire = datetime.now(timezone.utc) + expires_delta
        else:
            expire = datetime.now(timezone.utc) + timedelta(
                minutes=settings.access_token_expire_minutes
            )

        to_encode = {
            "sub": user_id,
            "exp": expire,
        }
        return jwt.encode(to_encode, settings.secret_key, algorithm=settings.algorithm)

    @staticmethod
    def decode_access_token(token: str) -> Optional[str]:
        """解码JWT令牌，返回用户ID"""
        try:
            payload = jwt.decode(
                token, settings.secret_key, algorithms=[settings.algorithm]
            )
            user_id: str = payload.get("sub")
            if user_id is None:
                return None
            return user_id
        except JWTError:
            return None

File: backend/src/domain/services/authService.ts
import bcrypt from 'bcrypt';
import jwt from 'jsonwebtoken';
import { UserCreate, UserResponse, TokenResponse } from '../entities/user';
import { UserModel } from '../../infrastructure/persistence/mysql/models/user';
import { settings } from '../../config';

export class AuthService {
  public async register(userData: UserCreate): Promise<UserResponse> {
    // Check if user already exists
    const existingUser = await UserModel.findOne({
      where: { email: userData.email },
    });

    if (existingUser) {
      throw new Error('User with this email already exists');
    }

    // Hash password
    const hashedPassword = await bcrypt.hash(userData.password, 10);

    // Create user
    const user = await UserModel.create({
      username: userData.username,
      email: userData.email,
      password: hashedPassword,
    });

    return this.mapToUserResponse(user);
  }

  public async login(email: string, password: string): Promise<TokenResponse> {
    // Find user
    const user = await UserModel.findOne({
      where: { email },
    });

    if (!user) {
      throw new Error('Invalid email or password');
    }

    // Verify password
    const isPasswordValid = await bcrypt.compare(password, user.password);

    if (!isPasswordValid) {
      throw new Error('Invalid email or password');
    }

    // Generate token
    const token = this.generateToken(user.id);

    return {
      access_token: token,
      token_type: 'bearer',
      user: this.mapToUserResponse(user),
    };
  }

  public async getUserById(id: string): Promise<UserResponse | null> {
    const user = await UserModel.findByPk(id);

    if (!user) {
      return null;
    }

    return this.mapToUserResponse(user);
  }

  private generateToken(userId: string): string {
    const payload = {
      sub: userId,
      iat: Math.floor(Date.now() / 1000),
      exp: Math.floor(Date.now() / 1000) + settings.accessTokenExpireMinutes * 60,
    };

    return jwt.sign(payload, settings.secretKey, {
      algorithm: settings.algorithm as jwt.Algorithm,
    });
  }

  public verifyToken(token: string): string {
    try {
      const payload = jwt.verify(token, settings.secretKey, {
        algorithms: [settings.algorithm as jwt.Algorithm],
      }) as any;

      return payload.sub;
    } catch (error) {
      throw new Error('Invalid or expired token');
    }
  }

  private mapToUserResponse(user: UserModel): UserResponse {
    return {
      id: user.id,
      username: user.username,
      email: user.email,
      created_at: user.created_at,
      updated_at: user.updated_at,
    };
  }
}
